### PR TITLE
chore: fixed a bug where formview's are not scrollable when the form is bigger than the screen height and the keyboard would cover most of the form

### PR DESCRIPTION
**Pull request by mohammedDehairy**
chore: fixed a bug where formview's are not scrollable when the form is bigger than the screen height and the keyboard would cover most of the form.

### CODE REVIEW

File: Adyen/Utilities/KeyboardObserver.swift [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+    /// :nodoc:
+    func subscribeToKeyboardUpdates(_ observer: @escaping (_ keyboardRect: CGRect) -> Void) -> Any
+
+    /// :nodoc:
+    func removeObserver(_ observer: Any)
+}
+
+/// :nodoc:
+extension KeyboardObserver {
```
Comment (descorp) [inline Adyen/Utilities/KeyboardObserver.swift:20]: Is this wise to hide `addObserver` behind protocol?
You need to delete observers on `deinit` and it is easy to forget if you didn't subscribed with your own hand..

File: AdyenDropIn/Presentation/WrapperViewController.swift [hunk 2]
```diff
@@ -48,6 +51,19 @@ internal final class WrapperViewController: UIViewController {
                        completion: nil)
     }
 
+    private func updateTopScrollViewInsets(keyboardHeight: CGFloat,
+                                           preferredContentSize: CGSize,
+                                           finalHeight: CGFloat) {
+        guard keyboardHeight > 0 else { return }
```
Comment (descorp) [inline AdyenDropIn/Presentation/WrapperViewController.swift:57]: What if you dismissed keyboard and want to go back to normal?

**Comment by descorp** [comment]
What about FormViewController? It also have bottom constraint based of keyboard..

File: AdyenDropIn/Presentation/DropInNavigationController.swift [hunk 3]
```diff
@@ -24,11 +24,16 @@ internal final class DropInNavigationController: UINavigationController {
...
-        subscribeToKeyboardUpdates()
+        keyboardObserver = subscribeToKeyboardUpdates { [weak self] in
+            self?.keyboardRect = $0
+            self?.updateTopViewControllerIfNeeded()
+        }
     }
     
     deinit {
-        notificationCenter.removeObserver(self)
+        if let keyboardObserver = keyboardObserver {
```
[the remark this answers is not in the document]
Comment (mohammedDehairy) [reply AdyenDropIn/Presentation/DropInNavigationController.swift]: very nice suggestion, thanks.

I prefer `keyboardObserver.map(removeObserver(_:))`, but don't have strong opinion :-).

File: AdyenDropIn/Presentation/WrapperViewController.swift [hunk 4]
```diff
@@ -48,6 +51,19 @@ internal final class WrapperViewController: UIViewController {
...
+        let bottomScrollInset: CGFloat = preferredContentSize.height + keyboardHeight - finalHeight
+        topMostScrollView?.contentInset.bottom = bottomScrollInset
+    }
+
+    private lazy var topMostScrollView: UIScrollView? = {
```
Comment (descorp) [inline AdyenDropIn/Presentation/WrapperViewController.swift:65]: Maybe we shouldn't look for UIScrollView, but rather create protocol for "ScrollingContentViewController" or something and look for it?

File: Adyen/Helpers/UIViewHelpers.swift [hunk 5]
```diff
@@ -19,4 +20,20 @@ extension AdyenScope where Base: UIView {
 
         return image
     }
+
+    /// :nodoc:
+    public func getTopMostView<T: UIView>() -> T? {
+        var queue = [UIView]()
```
Comment (kaphacius) [inline Adyen/Helpers/UIViewHelpers.swift:26]: can be simplified as `var queue = [base]`

File: AdyenDropIn/Presentation/WrapperViewController.swift (hunk 4 above)
Comment (mohammedDehairy) [reply AdyenDropIn/Presentation/WrapperViewController.swift]: yeah, just did something similar.

File: Adyen/Helpers/UIViewHelpers.swift (hunk 5 above)
Comment (mohammedDehairy) [reply Adyen/Helpers/UIViewHelpers.swift]: 👍

File: Adyen/Helpers/UIViewHelpers.swift [hunk 6]
```diff
@@ -19,4 +20,19 @@ extension AdyenScope where Base: UIView {
 
         return image
     }
+
+    /// :nodoc:
+    public func getTopMostView<T>() -> T? {
```
Comment (descorp) [inline Adyen/Helpers/UIViewHelpers.swift:25]: This is breadth search, not deep search.. `TopMost` is confusing naming 🤓

Comment (mohammedDehairy) [reply Adyen/Helpers/UIViewHelpers.swift:25]: well, breadth search will get the top most view, but depth first search will return the deepest view first, hence the `topMost` in the name, why is it confusing?.

Comment (descorp) [reply Adyen/Helpers/UIViewHelpers.swift:25]: If you have (this is just a weird en example 😅):

```
[UIView1] - [UIStackView] - [UIView3] - [UISctollView3] - [UIView4]
               |
           [UISctollView2] - [UIView5]
```

So UISctollView3 is "topMost", but this method will return UISctollView2.

It is not a problem or something, just saying the name is not describing what it does :D 
Should be `getFirstViewOfType<T>` or change `queue.removeFirst()` -> `queue.removeLast()`

File: AdyenDropIn/Presentation/WrapperViewController.swift [hunk 7]
```diff
@@ -48,6 +51,22 @@ internal final class WrapperViewController: UIViewController {
...
+                                           finalHeight: CGFloat) {
+        guard keyboardHeight > 0 else {
+            topMostScrollView?.contentInset.bottom = 0
+            return
+        }
+        let bottomScrollInset: CGFloat = preferredContentSize.height + keyboardHeight - finalHeight
+        topMostScrollView?.contentInset.bottom = bottomScrollInset
+    }
+
+    private lazy var topMostScrollView: KeyboardAvoidingView? = {
```
Comment (kaphacius) [inline AdyenDropIn/Presentation/WrapperViewController.swift:65]: can me simplified as `child.viewIfLoaded.flatMap { $0.adyen.getTopMostView() }`
no `return` needed :)

Comment (mohammedDehairy) [reply AdyenDropIn/Presentation/WrapperViewController.swift:65]: 👍

**Review by descorp — APPROVED** [review]

**Review by kaphacius — APPROVED** [review]

**Pull request merged**